Add index codec documentation (#4589)
opensearch-project/documentation-website: _im-plugin/index-settings.md

index.codec.compression_level | The compression level setting provides a tradeoff between compression ratio and speed. A higher compression level results in a higher compression ratio (smaller storage size) with a tradeoff in speed (slower compression and decompression speeds lead to greater indexing and search latencies). Can only be specified if `index.codec` is set to `zstd` and `zstd_no_dict` compression levels in OpenSearch 2.9 and later. Valid values are integers in the [1, 6] range. For more information, see [Index codec settings]({{site.url}}{{site.baseurl}}/im-plugin/index-codecs/). Optional. Default is 3. 

> index.routing_partition_size | The number of shards a custom routing value can go to. Routing helps an imbalanced cluster by relocating values to a subset of shards rather than a single shard. To enable, set this value to greater than 1 but less than `index.number_of_shards`. Default is 1.

Reviewer: Let's add a noun or pronoun after "To enable". To enable what?

index.soft_deletes.retention_lease.period | The maximum amount of time to retain a shard's history of operations. Default is `12h`.

> index.load_fixed_bitset_filters_eagerly | Whether OpenSearch should pre-load cached filters. Available options are `true` and `false`. Default is `true`.

Reviewer: ```suggestion
index.load_fixed_bitset_filters_eagerly | Whether OpenSearch should preload cached filters. Available options are `true` and `false`. Default is `true`.
```

## Updating a static index setting

You can update a static index setting only on a closed index. The following example demonstrates updating the index codec setting.

First, close an index: 

```json
POST /testindex/_close
```
{% include copy-curl.html %}

Then update the settings by sending a request to the `_settings` endpoint:

```json
PUT /testindex/_settings
{
  "index": {
    "codec": "zstd_no_dict",
    "codec.compression_level": 3
  }
}
```
{% include copy-curl.html %}

Last, reopen the index to enable read and write operations:

```json
POST /testindex/_open
```
{% include copy-curl.html %}

For more information about updating settings, including supported query parameters, see [Update settings]({{site.url}}{{site.baseurl}}/api-reference/index-apis/update-settings/).

## Dynamic index settings

The following table lists all available dynamic index settings.

Setting | Description
:--- | :---
index.number_of_replicas | The number of replica shards each primary shard should have. For example, if you have 4 primary shards and set `index.number_of_replicas` to 3, the index has 12 replica shards. Default is 1.

> index.auto_expand_replicas | Whether the cluster should automatically add replica shards based on the number of data nodes. Specify a lower bound and upper limit (for example, 0--9), or `all` for the upper limit. For example, if you have 5 data nodes and set `index.auto_expand_replicas` to 0--3, then the cluster does not automatically add another replica shard. However, if you set this value to `0-all` and add 2 more nodes for a total of 7, the cluster will expand to now have 6 replica shards. Default is disabled.

Reviewer: ```suggestion
index.auto_expand_replicas | Whether the cluster should automatically add replica shards based on the number of data nodes. Specify a lower bound and upper limit (for example, 0--9) or `all` for the upper limit. For example, if you have 5 data nodes and set `index.auto_expand_replicas` to 0--3, then the cluster does not automatically add another replica shard. However, if you set this value to `0-all` and add 2 more nodes for a total of 7, the cluster will expand to now have 6 replica shards. Default is disabled.
```

> index.search.idle.after | Amount of time a shard should wait for a search or get request until it goes idle. Default is `30s`.

Reviewer: ```suggestion
index.search.idle.after | The amount of time a shard should wait for a search or get request until it goes idle. Default is `30s`.
```

index.refresh_interval | How often the index should refresh, which publishes its most recent changes and makes them available for searching. Can be set to `-1` to disable refreshing. Default is `1s`.

> index.max_result_window | The maximum value of `from` + `size` for searches to the index. `from` is the starting index to search from, and `size` is the amount of results to return. Default: 10000.

Reviewer 2: ```suggestion
index.max_result_window | The maximum value of `from` + `size` for searches to the index. `from` is the starting index to search from, and `size` is the amount of results to return. Default is 10000.
```

> index.max_result_window | The maximum value of `from` + `size` for searches to the index. `from` is the starting index to search from, and `size` is the amount of results to return. Default is 10000.

Reviewer: ```suggestion
index.max_result_window | The maximum value of `from` + `size` for searches of the index. `from` is the starting index to search from, and `size` is the number of results to return. Default is 10000.
```

> index.max_inner_result_window | Maximum value of `from` + `size` to return nested search hits and most relevant document aggregated during the query. `from` is the starting index to search from, and `size` is the amount of top hits to return. Default is 100.

Reviewer: ```suggestion
index.max_inner_result_window | The maximum value of `from` + `size` to return nested search hits and most relevant document aggregated during the query. `from` is the starting index to search from, and `size` is the number of top hits to return. Default is 100.
```

> index.max_inner_result_window | Maximum value of `from` + `size` to return nested search hits and most relevant document aggregated during the query. `from` is the starting index to search from, and `size` is the amount of top hits to return. Default is 100.

Reviewer: The first sentence here isn't making sense to me. Can we rephrase/clarify?

> index.max_rescore_window | The maximum value of `window_size` for rescore requests to the index. Rescore requests reorder the index's documents and return a new score, which can be more precise. Default is the same as index.max_inner_result_window or 10000 by default.

Reviewer: Are there missing tic marks in here?

> index.max_docvalue_fields_search | Maximum amount of `docvalue_fields` allowed in a query. Default is 100.

Reviewer: ```suggestion
index.max_docvalue_fields_search | The maximum number of `docvalue_fields` allowed in a query. Default is 100.
```

> index.max_script_fields | Maximum amount of `script_fields` allowed in a query. Default is 32.

Reviewer: ```suggestion
index.max_script_fields | The maximum number of `script_fields` allowed in a query. Default is 32.
```

> index.max_ngram_diff | Maximum difference between `min_gram` and `max_gram` values for `NGramTokenizer` and `NGramTokenFilter` fields. Default is 1.

Reviewer: ```suggestion
index.max_ngram_diff | The maximum difference between `min_gram` and `max_gram` values for `NGramTokenizer` and `NGramTokenFilter` fields. Default is 1.
```

> index.max_ngram_diff | Maximum difference between `min_gram` and `max_gram` values for `NGramTokenizer` and `NGramTokenFilter` fields. Default is 1.

Reviewer: Should "the" precede `min_gram` and `NGramTokenizer`?

> index.max_shingle_diff | Maximum difference between `max_shingle_size` and `min_shingle_size` to feed into the `shingle` token filter. Default is 3.

Reviewer: ```suggestion
index.max_shingle_diff | The maximum difference between `max_shingle_size` and `min_shingle_size` to feed into the `shingle` token filter. Default is 3.
```

> index.max_refresh_listeners | Maximum amount of refresh listeners each shard is allowed to have.

Reviewer: ```suggestion
index.max_refresh_listeners | The maximum number of refresh listeners each shard is allowed to have.
```

> index.analyze.max_token_count | Maximum amount of tokens that can return from the `_analyze` API operation. Default is 10000.

Reviewer: ```suggestion
index.analyze.max_token_count | The maximum number of tokens that can return from the `_analyze` API operation. Default is 10000.
```

> index.analyze.max_token_count | Maximum amount of tokens that can return from the `_analyze` API operation. Default is 10000.

Reviewer: "be returned" instead of "return"?

> index.highlight.max_analyzed_offset | The amount of characters a highlight request can analyze. Default is 1000000.

Reviewer: ```suggestion
index.highlight.max_analyzed_offset | The number of characters a highlight request can analyze. Default is 1000000.
```

> index.max_terms_count | The maximum amount of terms a terms query can accept. Default is 65536.

Reviewer: ```suggestion
index.max_terms_count | The maximum number of terms a terms query can accept. Default is 65536.
```

index.max_regex_length | The maximum character length of regex that can be in a regexp query. Default is 1000.
index.query.default_field | A field or list of fields that OpenSearch uses in queries in case a field isn't specified in the parameters.

> index.routing.allocation.enable | Specifies options for the index’s shard allocation. Available options are all (allow allocation for all shards), primaries (allow allocation only for primary shards), new_primaries (allow allocation only for new primary shards), and none (do not allow allocation). Default is all.

Reviewer: Are there missing tic marks here?

index.routing.rebalance.enable | Enables shard rebalancing for the index. Available options are `all` (allow rebalancing for all shards), `primaries` (allow rebalancing only for primary shards), `replicas` (allow rebalancing only for replicas), and `none` (do not allow rebalancing). Default is `all`.

> index.gc_deletes | Amount of time to retain a deleted document's version number. Default is `60s`.

Reviewer: ```suggestion
index.gc_deletes | The amount of time to retain a deleted document's version number. Default is `60s`.
```

index.default_pipeline | The default ingest node pipeline for the index. If the default pipeline is set and the pipeline does not exist, then index requests fail. The pipeline name `_none` specifies that the index does not have an ingest pipeline.
index.final_pipeline | The final ingest node pipeline for the index. If the final pipeline is set and the pipeline does not exist, then index requests fail. The pipeline name `_none` specifies that the index does not have an ingest pipeline.

## Updating a dynamic index setting

> You can update a dynamic index setting any time through the API. For example, to update the refresh interval, use the following request:

Reviewer: ```suggestion
You can update a dynamic index setting at any time through the API. For example, to update the refresh interval, use the following request:
```
